### PR TITLE
Use more efficient row/col computation

### LINKED ISSUE #151351: Sparse tensor conversion performance issues (CPU/GPU)
### 🐛 Describe the bug

I was investigating opportunities for improving "activation" sparsity as follows

```
import time
import torch

d = 1024 * 32
A = -torch.ones(d, d)
A[0, 0] = 111
A[10, 10] = 222     # most entries in A are < 0

T = torch.relu(A)         # materializes very sparse T as dense first
S = T.to_sparse_csr()     # would be nice to have something like S = torch.sparse_relu(A)
                          # but that is not the point of this bug yet
```

where I really would like to have a "sparsifying" relu to avoid materializing T as dense intermediate first.

However, while pondering on that, I noticed a huge performance (time and memory) difference between converting to COO or converting to CSR. Take the following annotated code

```
.. construct A..
time.sleep(10)      # INTERVAL 1
T = torch.relu(A)   # materializes very sparse T as dense first
time.sleep(10)      # INTERVAL 2
S = T.to_sparse()   # to COO
time.sleep(10)      # INTERVAL 3
```

and compare memory tracking for COO (to_sparse) with CSR (to_sparse_csr). Attached are the two plots. Both are running on CPU (but the problem occurs for GPU as well, same code path).

It is clear that COO behaves more or less as expected (first bump to get A, second bump to get T, then nothing more for S; again, my initial goal was to avoid the second bump for T, but read on). Then looking at CSR, we get a similar first bump to get A, second bump to T, but then a huge increase in memory to get to S. Also, the time increases substantially (notice the extra time in between INTERVAL 2 and INTERVAL 3).

Image

Image

### Versions

all versions

cc @alexsamardzic @nikitaved @pearu @cpuhrsch @amjames @bhosmer @jcaip

### PR DESCRIPTION
**Pull request by aartbik**
This change addresses the first/second time/mem "spike" observed in

https://github.com/pytorch/pytorch/issues/151351

Fixes #151351

### CODE REVIEW

**Review by eqy — APPROVED** [review]

**Review by amjames — APPROVED** [review]
More performant and less code LGTM.

**Review by Skylion007 — APPROVED** [review]

**Comment by Skylion007** [comment]
Actually @aartbik , I was thinking, could these two PRs be simplified even further by using the `nonzero(as_tuple=True)` arg in C++?

File: aten/src/ATen/native/TensorConversions.cpp [hunk 1]
```diff
@@ -1026,19 +1026,8 @@ static std::pair<Tensor, Tensor> _not_zero_mask_to_col_row_indices(
...
-          .masked_select(not_zero_mask);
-  auto row_indices =
-      at::native::arange(
-          not_zero_mask.size(-2), index_dtype, kStrided, index_device)
-          .view({not_zero_mask.size(-2), 1})
-          .expand_as(not_zero_mask)
-          .masked_select(not_zero_mask);
-  return std::pair<Tensor, Tensor>(col_indices, row_indices);
+  auto nz = not_zero_mask.nonzero().transpose(0, 1);
+  return std::pair<Tensor, Tensor>(nz[1], nz[0]);
```
Comment (Skylion007) [inline aten/src/ATen/native/TensorConversions.cpp:1030]:
```suggestion
  auto [rows, cols] = not_zero_mask.nonzero(/*as_tuple=*/true);
  return {std::move(cols), std::move(rows)};
```

File: aten/src/ATen/native/TensorConversions.cpp [hunk 2]
```diff
@@ -1026,19 +1026,8 @@ static std::pair<Tensor, Tensor> _not_zero_mask_to_col_row_indices(
     Tensor not_zero_mask,
     ScalarType index_dtype,
     Device index_device) {
-  auto col_indices =
-      at::native::arange(
-          not_zero_mask.size(-1), index_dtype, kStrided, index_device)
```
Comment (Skylion007) [inline aten/src/ATen/native/TensorConversions.cpp:1031]: All these were the last 2 indices here... Are we sure the dimensional of the shape is right on higher dim sparse inputs?

**Comment by aartbik** [comment]
@Skylion007 I had similar concerns about higher dims as well and pondered over adding "TORCH_CHECK(not_zero_mask.dim() == 2" to make sure we are not encountering those cases (but it seems that this feature is not tested at all, and I could not construct cases that broke the current code?)

Of course feel free to replace my code with better solutions, removing the "densifying" bottleneck was really my only objective, since it hampers activation sparsity cases quite a bit.
